### PR TITLE
(FACT-2481) Fix caching

### PR DESCRIPTION
**Pull request by BogdanIrimie**

### CODE REVIEW

File: lib/resolvers/macosx/system_profiler_resolver.rb [hunk 1]
```diff
@@ -30,14 +30,13 @@ class SystemProfiler < BaseResolver
 
       @log = Facter::Log.new(self)
       @semaphore = Mutex.new
+      @fact_list = {}
```
Comment (Filipovici-Andrei) [inline lib/resolvers/macosx/system_profiler_resolver.rb]: In other resolvers you have @fact_list ||= {}

File: lib/resolvers/macosx/processor_resolver.rb [hunk 2]
```diff
@@ -6,7 +6,7 @@ module Macosx
       class Processors < BaseResolver
         @log = Facter::Log.new(self)
         @semaphore = Mutex.new
-        @fact_list ||= {}
+        @fact_list = {}
```
Comment (Filipovici-Andrei) [inline lib/resolvers/macosx/processor_resolver.rb]: In other resolvers you have @fact_list ||= {}

**Review by Filipovici-Andrei — CHANGES_REQUESTED** [review]
I think we are going in the wrong direction with the @fact_list and @semaphore. They shouldn't be tampered with in the resolvers. They should exist only in the BaseResolver. BaseResolver should be like a template, for what resolvers should do, and each resolver would implement just the post_resolve method. Caching, logging and synchronising should be dealt with in just ONE PLACE, in the BaseResolver.

File: lib/resolvers/macosx/processor_resolver.rb (hunk 2 above)
Comment (IrimieBogdan) [reply lib/resolvers/macosx/processor_resolver.rb]: As `@fact_list` is initialise on class creation, there is no need for `||=`

File: lib/resolvers/macosx/system_profiler_resolver.rb (hunk 1 above)
Comment (IrimieBogdan) [reply lib/resolvers/macosx/system_profiler_resolver.rb]: As `@fact_list` is initialise on class creation, there is no need for `||=`

File: lib/resolvers/networking_linux_resolver.rb [hunk 3]
```diff
@@ -4,12 +4,13 @@ module Facter
   module Resolvers
     class NetworkingLinux < BaseResolver
       @semaphore = Mutex.new
+      @fact_list ||= {}
```
Comment (gimmyxd) [inline lib/resolvers/networking_linux_resolver.rb]: should this be `@fact_list = {}` as in other resolvers?

**Comment by IrimieBogdan** [comment]
@Filipovici-Andrei `@semaphore` serialize access to one resolver, it should be declared and used at a resolver level, otherwise it would serialize access to all resolvers! Caching is done at the resolver level and should be kept at this level because serialisation is done at this level. The logger can be moved to base resolver.

File: lib/resolvers/networking_linux_resolver.rb (hunk 3 above)
Comment (IrimieBogdan) [reply lib/resolvers/networking_linux_resolver.rb]: Yes

**Review by Filipovici-Andrei — DISMISSED** [review]

**Review by gimmyxd — APPROVED** [review]

**Pull request merged**